world_positioning/src/worldP_listener.py

```diff
@@ -0,0 +1,206 @@
+                                    msg.pos_variance.y,
+                                    msg.pos_variance.z])
+        self.pos_var.shape = (3,1)
+        #input velocity
+        self.velocity = np.array(  [msg.velocity.x,
+                                    msg.velocity.y,
+                                    msg.velocity.z])[np.newaxis]
+        self.velocity.shape = (3,1)
+        #input velocity varience
+        self.vel_var = np.array(   [msg.vel_variance.x,
+                                    msg.vel_variance.y,
+                                    msg.vel_variance.z])
+        self.vel_var.shape = (3,1)
+        #input orientation
+        self.orientation = np.array(   [msg.orientation_RPY.w, ##not sure whether rpy is the right one
+                                        msg.orientation_RPY.x,
+                                        msg.orientation_RPY.y,
+                                        msg.orientation_RPY.z])[np.newaxis]
+        self.orientation.shape = (4,1)
+        #input orientation varience
+        self.orient_var = np.array(    [msg.orientation_RPY_variance.w, ##not sure whether rpy is the right one
+                                        msg.orientation_RPY_variance.x,
+                                        msg.orientation_RPY_variance.y,
+                                        msg.orientation_RPY_variance.z])
+        self.orient_var.shape = (4,1)
+        #input angular velocity
+        self.angular_vel = np.array(   [msg.ang_vel.x,
+                                        msg.ang_vel.y,
+                                        msg.ang_vel.z])[np.newaxis]
+        self.angular_vel.shape = (3,1)
+        #input angular velocity varience
+        self.angular_vel_var = np.array(   [msg.ang_vel_varience.x,
+                                            msg.ang_vel_varience.y,
+                                            msg.ang_vel_varience.z])
+        self.angular_vel_var.shape = (3,1)
+
+        self.last_time = rospy.Time.now().to_sec()
+
+class ZedCamListener(BaseListener):
+    def __init__(self):
+        super(SubStateListener, self).__init__()
+        #currently working for a single object
+        self.cam_sub = rospy.Subscriber("zed_camera",Point,self.cam_cb)
+        
+        self.obj_pos = np.array([0,0,0])
+        self.obj_pos_var = np.array([0,0,0])
+
+    def get_timeout_sec(self):
+        return 0.1
+    
+    def get_p(self):
+        #single point, 3 elements
+        return 3
+
+    def get_measurement_z(self):
+        vec = np.zeros((self.get_p(),1))
+        for i in range(0,3):
+            vec[i] = self.obj.pos[i]
+        return vec
+    
+    def get_R(self):
+        var = np.zeros((self.get_p(),1))
+        for i in range(0,3):
+            var[i,i] = self.obj_pos_var[i]
+        return var
+
+    def state_to_measurement_h(self, x, u):
+        #currently 3 element point
+        z = np.array([x[13],x[14],x[15]])  # from substate + 3 points
```

Comment:
not sure if this is ok, i currently have the state_to_measurement_h just returning the same camera position of the object, and the x will just contain the raw camera position of each object. I'm doing the processing of camera position to absolute position in the construction of the world position message. I'm trying to get it so that x has relative object positions instead of raw camera positions, but if using the raw camera positions is fine, then i'm just going to leav eit.

Reply:
I think making x have global positions of objects is the best bet. If x stores the relative location, we would need to update it for each object whenever the sub moves, and I think that would add too much noise.

Something like:
  x: global positions of all objects
  incoming message: has relative location of object as seen by camera
  z function: returns the relative measurement
  state_to_measurement_h: calculates what the relative measurement should be given the two absolution positions and sub orientation

Reply:
While I was working on that, I was considering how to deal with the variance of the object positions, cuz they're still coming out of the get_R function as relative. They're supposed to be kinda small though, so would it be ok if they just stayed relative until the data published as a message? Otherwise, I could try changing the get_R function to also take x and u so i can calculate the absolute values.

Reply:
The get_R function should be returning the uncertainty of the measurements (relative position), not the state (absolute position).